server/src/main/java/org/opensearch/rest/NamedRoute.java

```diff
@@ -9,47 +9,168 @@
+
+        /**
+         * Sets the handler for this route
+         *
+         * @param handler the handler for this route
+         * @return the builder instance
+         */
+        public Builder handler(Function<RestRequest, RestResponse> handler) {
+            if (handler == null) {
+                throw new OpenSearchException("Route handler must not be null.");
+            }
+            this.handler = handler;
+            return this;
         }
-        return routeName.matches(VALID_ACTION_NAME_PATTERN);
+
+        /**
+         * Builds a new instance of {@link NamedRoute} based on the provided parameters.
+         *
+         * @return a new instance of {@link NamedRoute}
+         * @throws OpenSearchException if the route name is invalid
+         */
+        public NamedRoute build() {
+            return new NamedRoute(this);
+        }
+
+        private Set<String> validateLegacyActionNames(Set<String> legacyActionNames) {
+            if (legacyActionNames == null) {
+                return new HashSet<>();
+            }
+            for (String actionName : legacyActionNames) {
+                if (!TransportService.isValidActionName(actionName)) {
+                    throw new OpenSearchException(
+                        "Invalid action name [" + actionName + "]. It must start with one of: " + TransportService.VALID_ACTION_PREFIXES
+                    );
+                }
+            }
+            return legacyActionNames;
+        }
+
     }
 
-    public NamedRoute(RestRequest.Method method, String path, String name) {
-        super(method, path);
-        if (!isValidRouteName(name)) {
+    private NamedRoute(Builder builder) {
+        super(builder.method, builder.path);
+        if (!isValidRouteName(builder.uniqueName)) {
             throw new OpenSearchException(
                 "Invalid route name specified. The route name may include the following characters"
                     + " 'a-z', 'A-Z', '0-9', ':', '/', '*', '_' and be less than "
                     + MAX_LENGTH_OF_ACTION_NAME
                     + " characters"
             );
         }
-        this.name = name;
+        this.uniqueName = builder.uniqueName;
+        this.actionNames = builder.legacyActionNames;
+        this.handler = builder.handler;
+    }
+
+    public boolean isValidRouteName(String routeName) {
+        return routeName != null
+            && !routeName.isBlank()
+            && routeName.length() <= MAX_LENGTH_OF_ACTION_NAME
+            && routeName.matches(VALID_ACTION_NAME_PATTERN);
     }
 
     /**
      * The name of the Route. Must be unique across Route.
      */
     public String name() {
-        return this.name;
+        return this.uniqueName;
+    }
+
+    /**
+     * The legacy transport Action name to match against this route to support authorization in REST layer.
+     * MUST be unique across all Routes
+     */
+    public Set<String> actionNames() {
+        return this.actionNames;
```

Comment:
This returns the mutable set directly unless you've already made it unmodifiable per my previous comment. If someone later added to the action name set in the builder it might update the returned value if it's used anywhere. 

If you didn't use `Set.copyOf()` when setting `actionNames` you need to do so here (but it should be done earlier).